### PR TITLE
[Form] Deprecated usage of "choices" option in sub types

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #... 
| License       | MIT
| Doc PR        | ~

Follows #21880 and the discussion in #20771.

### CODE REVIEW

File: UPGRADE-4.0.md
Comment (xabbuh) [inline UPGRADE-4.0.md]: Should we really support both ways to solve it?

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: Either we should only support the `choice_loader` solution in the future or talk about the possibility to explicitly set it to `null` here too.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ExtendedChoiceTypeTest.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ExtendedChoiceTypeTest.php]: no need to change the name here

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: I hesitated, considering your comments I'd rather add it explicitly here too, thanks.

File: UPGRADE-4.0.md
Comment (HeahDude) [inline UPGRADE-4.0.md]: Yes I think so, if one wants to override choices, he should not be force to override the loader.

It makes no sense IMO to enforce lazy load something that is already loaded, if you have your choices, just pass them, do not encapsulate them in an objet that will eventually call a function to return the initial thing.

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: Done, I've just updated the notices.

File: UPGRADE-4.0.md
Comment (stof) [inline UPGRADE-4.0.md]: @xabbuh supporting a single way to achieve it would actually involve **more** work to forbid one of them. Both ways are just features of the ChoiceType: if ``choice_loader`` is not null, it is used. Otherwise ``choices`` is used.

The CountryType & co currently have to do extra work to ensure that this priority order is reversed, due to BC.

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: Using the "choices" option [...]

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ExtendedChoiceTypeTest.php
Comment (xabbuh) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ExtendedChoiceTypeTest.php]: Do we really need the type extension here?

Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ExtendedChoiceTypeTest.php]: I would keep it like it for consistency with the test above, and because it test both overriding the option directly and doing it with an extension, so looks safer as is.

File: src/Symfony/Component/Form/Extension/Core/Type/CountryType.php
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/CountryType.php]: Fixed everywhere.

**Comment by HeahDude** [comment]
Comments addressed and rebased. Ready for final review before merging, ping @symfony/deciders.

File: UPGRADE-3.3.md
Comment (xabbuh) [inline UPGRADE-3.3.md]: I think we need to reword this a bit to make clear that this is only deprecated as long as you do not set the `choice_loader` option to `null` (something like "using the `choices` option without overriding the `choice_loader` option is deprecated").

Comment (HeahDude) [inline UPGRADE-3.3.md]: Good idea! Done.

**Review by xabbuh — APPROVED** [review]
:+1:

File: UPGRADE-3.3.md
Comment (Tobion) [inline UPGRADE-3.3.md]: That doesn't look intuitive to me.

Comment (stof) [inline UPGRADE-3.3.md]: Well, this is because choice_loader has priority over choices in ChoiceType (which is not documented IIRC btw)

Comment (HeahDude) [inline UPGRADE-3.3.md]: And this should not be used ideally, the ChoiceType is more accurate in such cases, I'll update the doc to clarify this.

Comment (fabpot) [inline UPGRADE-3.3.md]: Wouldn't it be "better" to drop this example and just keep the second code snippet? And remove the recommendation to use `choice_loader` as null?

Comment (HeahDude) [inline UPGRADE-3.3.md]: I understand this does not look pretty, but this is the actual quick fix for some breaking app.

What we can't do is to stop supporting "choices" option as it's inherited from the ChoiceType, and in such cases `null` is needed because of its precedence.

What we can do is to better document these types options, and explicit that "choices" or "choice_loader" options should be overridden on a ChoiceType or an EntityType only, not on a child already doing so, unless the previous value is reused.

I would like to address your comment, but as I said in a previous outdated diff:
>It makes no sense IMO to enforce lazy load something that is already loaded, if you have your choices, just pass them, do not encapsulate them in an objet that will eventually call a function to return the initial thing.

Then the real fix would be to use the ChoiceType (if the previous "choices" values is not used), I could add this instead, but objectively I would let it as is now.

I'm waiting for your final call on this symfony deciders.

Comment (xabbuh) [inline UPGRADE-3.3.md]: I am fine with how it is now. But we should indeed clarify the documentation about the differences of the `choices` and `choice_loader` options and when to use what.

**Comment by fabpot** [comment]
Thank you @HeahDude.

**Pull request merged**
